agent/probes.go

```diff
@@ -39,100 +39,72 @@ import (
 	"github.com/skydive-project/skydive/topology/probes/vpp"
 )
 
-// NewTopologyProbeBundleFromConfig creates a new topology probe.Bundle based on the configuration
-func NewTopologyProbeBundleFromConfig(g *graph.Graph, hostNode *graph.Node) (*probe.Bundle, error) {
+// NewTopologyProbeBundle creates a new topology probe.Bundle based on the configuration
+func NewTopologyProbeBundle(g *graph.Graph, hostNode *graph.Node) (*probe.Bundle, error) {
 	list := config.GetStringSlice("agent.topology.probes")
 	logging.GetLogger().Infof("Topology probes: %v", list)
 
-	probes := make(map[string]probe.Probe)
-	bundle := probe.NewBundle(probes)
+	var handler probe.Handler
+	var err error
+
+	bundle := probe.NewBundle()
+	ctx := tp.Context{
+		Logger:   logging.GetLogger(),
+		Config:   config.GetConfig(),
+		Graph:    g,
+		RootNode: hostNode,
+	}
 
-	var nsProbe *netns.Probe
 	if runtime.GOOS == "linux" {
-		nlProbe, err := netlink.NewProbe(g, hostNode)
+		nlHandler, err := new(netlink.ProbeHandler).Init(ctx, bundle)
 		if err != nil {
 			return nil, err
 		}
-		probes["netlink"] = nlProbe
+		bundle.AddHandler("netlink", nlHandler)
 
-		nsProbe, err = netns.NewProbe(g, hostNode, nlProbe)
+		nsHandler, err := new(netns.ProbeHandler).Init(ctx, bundle)
 		if err != nil {
 			return nil, err
 		}
-		probes["netns"] = nsProbe
+		bundle.AddHandler("netns", nsHandler)
 	}
 
 	for _, t := range list {
-		if _, ok := probes[t]; ok {
+		if bundle.GetHandler(t) != nil {
 			continue
 		}
 
 		switch t {
 		case "ovsdb":
-			addr := config.GetString("ovs.ovsdb")
-			enableStats := config.GetBool("ovs.enable_stats")
-			ovsProbe, err := ovsdb.NewProbeFromConfig(g, hostNode, addr, enableStats)
-			if err != nil {
-				return nil, fmt.Errorf("Failed to initialize OVS probe: %s", err)
-			}
-			probes[t] = ovsProbe
+			handler, err = new(ovsdb.Probe).Init(ctx, bundle)
 		case "lxd":
-			lxdURL := config.GetConfig().GetString("lxd.url")
-			lxdProbe, err := lxd.NewProbe(nsProbe, lxdURL)
-			if err != nil {
-				return nil, fmt.Errorf("Failed to initialize LXD probe: %s", err)
-			}
-			probes[t] = lxdProbe
+			handler, err = new(lxd.ProbeHandler).Init(ctx, bundle)
 		case "docker":
-			dockerURL := config.GetString("agent.topology.docker.url")
-			netnsRunPath := config.GetString("agent.topology.docker.netns.run_path")
-			dockerProbe, err := docker.NewProbe(nsProbe, dockerURL, netnsRunPath)
-			if err != nil {
-				return nil, fmt.Errorf("Failed to initialize Docker probe: %s", err)
-			}
-			probes[t] = dockerProbe
+			handler, err = new(docker.ProbeHandler).Init(ctx, bundle)
 		case "lldp":
-			interfaces := config.GetStringSlice("agent.topology.lldp.interfaces")
-			lldpProbe, err := lldp.NewProbe(g, hostNode, interfaces)
-			if err != nil {
-				return nil, fmt.Errorf("Failed to initialize LLDP probe: %s", err)
-			}
-			probes[t] = lldpProbe
+			handler, err = new(lldp.Probe).Init(ctx, bundle)
```

Comment:
why are some named `Probe` while others `ProbeHandler` ?

Reply:
ProbeHandler are creating real probe meaning really capturing information. As example NetNS detect namespace and start netlink within it.